### ISSUE TITLE
Page Meta Service does not emit for Category Pages not having a product list component

### ISSUE THREAD

**Issue by janwidmer** [SAP/spartacus#11828]
**Describe the bug**
By adding a category restriction, (Can be added in the Backoffice => WCMS => Restrictions), it is possible to render e.g. the `CategoryPageTemplate` instead of the `ProductListTemplate`. This is useful, if the Customer has so called "Root" Categories, where he does not want to directly show products, but e.g. Category Teasers for all sub categories of this "Root" Category.

This works so far, Spartacus correctly renders the `CategoryPageTemplate`. The problem is, that the `PageMetaService` does not emit in such a case.

**Tell us the version of Spartacus**
- Library version: 3.1.0

**To Reproduce**
1. Create a Category Page using the Page type `CategoryPage` and the Template `CategoryPageTemplate`
2. Add a Page title for that page, e.g. `Category with active Category Restriction`
3. Add a category restriction for a certain Category (e.g. `c/578` on the spartacus demo instance) within the Backoffice (WCMS => Restrictions) and choose the previously created Category Page
4. Navigate to the category page, where you added the restriction (e.g. `c/578` on the spartacus demo instance)
5. Notice, that the page title tag only renders `Example Storefront` instead of the correct page title

**Expected behaviour**
The page title tag rendered in the head area of the html renders the cms page title of the previously created category page

**Screenshots**
Category Restriction in Backoffice:
![Category Restriction](https://user-images.githubusercontent.com/3381820/113680565-4840e880-96c1-11eb-922c-1b343032c1de.jpg)

Category Page Markup:
![Category Page Markup](https://user-images.githubusercontent.com/3381820/113680594-50008d00-96c1-11eb-8dfa-8c8d7a1b0178.jpg)

**Desktop (please complete the following information):**
 - OS: all
 - Browser all
 - Version all

**Smartphone (please complete the following information):**
 - Device: all
 - OS: all
 - Browser all
 - Version all

**Additional context**
It seems like, the problem happens in the `resolve` method of the `PageMetaService`:
![PageMeta Service Method resolve](https://user-images.githubusercontent.com/3381820/113683160-2d23a800-96c4-11eb-8e48-2d1cefdbcdaf.jpg)

Since the `resolve` method does not emit, the `meta$` property of the service does also not emit in that case. My assumption is, that the `resolveTitle` method does not emit within the `CategoryPageMetaResolver` because it filters for the pagination to exist:

```
resolveTitle(): Observable<string> {
    return (<Observable<ProductSearchPage>>this.searchPage$).pipe(
      filter((page: ProductSearchPage) => !!page.pagination),
      switchMap((p: ProductSearchPage) =>
        this.translation.translate('pageMetaResolver.category.title', {
          count: p.pagination?.totalResults,
          query: p.breadcrumbs?.length
            ? p.breadcrumbs[0].facetValueName
            : undefined,
        })
      )
    );
  }
```

As `this.searchPage$` returns a `Observable<Page>` if the page does not have a product list component:

```
 protected searchPage$: Observable<
    ProductSearchPage | Page
  > = this.cms.getCurrentPage().pipe(
    filter((page) => Boolean(page)),
    switchMap((page: Page) =>
      // only the existence of a plp component tells us if products
      // are rendered or if this is an ordinary content page
      this.hasProductListComponent(page)
        ? this.productSearchService
            .getResults()
            .pipe(filter((result) => Boolean(result)))
        : of(page)
    )
  );
  ```

**Workaround**
As workaround, I have implemented the following in my CustomCategroyPageMetaResolver:

```
protected categoryPage$: Observable<[Page, ProductSearchPage]> = this.cms.getCurrentPage().pipe(
		filter((page) => Boolean(page)),
		switchMap((page: Page) =>
			this.productSearchService.getResults().pipe(
				filter((result) => Boolean(result)),
				map((productSearchPage) => [page, productSearchPage] as [Page, ProductSearchPage])
			)
		)
	);
```

It allows me to access both the `ProductSearchPage` and the `Page` and decide in the resolver methods, which information I want to consider.

**Comment by tobi-or-not-tobi** [comment]
@janwidmer This seems a similar issue/limitation as was reported at #5680. We've hardcoded the condition to know when to use page vs product search page data for the resolver methods, it's only that the current approach uses hardcoded templates to figure this out. We've been playing with to ideas to improve this area: 
1. make the templates configurable in spartacus. relatively easy to achieve, but still requires you to do something
2. evaluate the cms components of the page payload to understand the data. This might be accurate in most cases, as these components arent' customized often. 

To unblock customers now, you can override the `hasProductListComponent` method, would that work for you?

**Comment by tobi-or-not-tobi** [comment]
closing in favor of #5680

**Comment by janwidmer** [comment]
@tobi-or-not-tobi yes it looks like, its related to #5680, besides that we are using a different page template though, in the other ticket, just a different component is used as far as I understand. It causes the same issue, that the Page Meta is not emitting anymore.

I guess Option No. 2 would no work for our case, as we use a custom component to show category teasers of the Current category. In that case, the component evaluation would need to be based on some kind of config.

How about an option No. 3 and just moving the `hasProductListComponent` check from the `searchPage$` observable to the `resolveTitle` method? The observable could then be based on the normal `Page` data together with the `ProductSearchPage` data and the developer could decide on his own, what he wants to use in which case. And it would prevent the PageMeta from not being emitted because of a the missing `pagination` property

Please see my workaround, which is basically my Option No. 3:

Extend the `CategoryPageMetaResolver` by adding:
```
protected categoryPage$: Observable<[Page, ProductSearchPage]> = this.cms.getCurrentPage().pipe(
		filter((page) => Boolean(page)),
		switchMap((page: Page) =>
			this.productSearchService.getResults().pipe(
				filter((result) => Boolean(result)),
				map((productSearchPage) => [page, productSearchPage] as [Page, ProductSearchPage])
			)
		)
	);

resolveTitle(): Observable<string> {
		return this.categoryPage$.pipe(
			switchMap(([page, productSearchPage]) => {
				if (
					this.hasProductListComponent(page) &&
					productSearchPage.breadcrumbs &&
					productSearchPage.breadcrumbs?.length > 0
				) {
					return of(
						productSearchPage.breadcrumbs.find(
							(breadcrumb) => breadcrumb.facetValueCode === productSearchPage.categoryCode
						)?.facetValueName || ''
					);
				} else {
					return of(page.name || '');
				}
			})
		);
	}
```
